### ISSUE TITLE
qpy does not support `Discriminator` and `Kernel`

### ISSUE THREAD

**Issue by wshanks** [Qiskit/qiskit#10050]
### Environment

- **Qiskit Terra version**: c4f1b0bd0b3873bcc5807a3c55b7eeceedd18193
- **Python version**:  3.10
- **Operating system**:  Fedora Linux 37

### What is happening?

Trying to serialize with qpy a circuit with a pulse gate which includes an `Acquire` instruction with a `Discriminator` or `Kernel` fails. The error is:

    QpyError: "Object type '<class 'qiskit.pulse.configuration.Discriminator'>' is not supported in Value namespace."

and similar for `Kernel`.

Note this means that with qiskit-ibm-provider pulse gates with discriminators or kernels are not currently supported since circuits must be serialized with qpy.

### How can we reproduce the issue?

Here are reproducers for `Discriminator`:

```python
from io import BytesIO

import numpy as np

from qiskit import QuantumCircuit, pulse
from qiskit.qpy import dump


qc = QuantumCircuit(1, 1)
qc.measure(0, 0)

discriminator = pulse.Discriminator("hw_centroid")

with pulse.builder.build() as sched:
    pulse.builder.acquire(10, 0, pulse.MemorySlot(0), discriminator=discriminator
    
qc.add_calibration("measure", (0,), sched)
    
file_ = BytesIO()
dump(qc, file_)
```

and `Kernel`:

```python
from io import BytesIO

import numpy as np

from qiskit import QuantumCircuit, pulse
from qiskit.qpy import dump


qc = QuantumCircuit(1, 1)
qc.measure(0, 0)

kernel = pulse.Kernel("hw_qmfk", kernel={"real": np.zeros(10), "imag": np.zeros(10)}, bias=[0, 0])

with pulse.builder.build() as sched:
    pulse.builder.acquire(10, 0, pulse.MemorySlot(0), kernel=kernel)
    
qc.add_calibration("measure", (0,), sched)
    
file_ = BytesIO()
dump(qc, file_)
```

### What should happen?

`Discriminator` and `Kernel` should be serializable with qpy. Since they are part of the circuit, they are already JSON serializable to support the old qobj path.

**Comment by nkanazawa1989** [comment]
Thanks Will. As we discussed offline, we tend to remove them from acquire instruction arguments unless we receive user request.

https://github.com/Qiskit/qiskit-terra/blob/ff80b61fef132930e87d63732d730f5eb5f2ae74/qiskit/qpy/type_keys.py#L322-L324

Keeping them in the acquire instruction indicates we can write a single quantum circuit including different kernels and discriminators, e.g. mid-circuit measurement, which is likely unrealistic. Instead of adding QPY support, we should deprecate them and define alternative path to describe kernel and descriminator.

In OpenPulse grammar, capture instruction is sort of arbitrary, but they show nice example of defining descriminator.

```python
defcalgrammar "openpulse";

cal {
    // Use a boxcar function to generate IQ data from raw waveform
    extern boxcar(waveform input) -> complex[float[64]];
    // Use a linear discriminator to generate bits from IQ data
    extern discriminate(complex[float[64]] iq) -> bit;
    ...
}

defcal measure $0 -> bit {
    ...
    waveform raw_output = capture_v1(capture_frame, 16000dt);

    // Kernel and discriminate
    complex[float[32]] iq = boxcar(raw_output);
    bit result = discriminate(iq);

    return result;
}
```
https://openqasm.com/language/openpulse.html#capture-instruction

**Comment by junnaka51** [comment]
@wshanks @nkanazawa1989 
Hello 👋 If currently no one is working on this issue, may I try it?

**Comment by nkanazawa1989** [comment]
Yes, please go ahead! I talked again with @wshanks , and decided to add QPY support of kernel and discriminator rather than dropping them from the acquire instruction. Do you think you can still handle this issue? Feel free to reach me out if you need any help :)

I implemented QPY support of another instruction with this PR. Just FYI
https://github.com/Qiskit/qiskit-terra/pull/9890

**Comment by junnaka51** [comment]
@nkanazawa1989 Thank you so much! 🙇

> Do you think you can still handle this issue?

I will see it in more detail if I can do so 😄

**Comment by junnaka51** [comment]
@nkanazawa1989 
Hello 👋 
The problem seems resolved from the observation that the error no longer raises in https://github.com/Qiskit/qiskit-terra/pull/10327

I would like to know whether this approach of mine appropriately follows the convention and/or the rule of `qpy` or not, before I update docs and add the release note. Could you please briefly check it and give me comments back? 🙏 
Thank you so much in advance! 🙇
